### ISSUE TITLE
Unary plus is not type checked

### ISSUE THREAD

**Issue by vkurchatkin** [facebook/flow#2643]
It's weird that Flows allows this:

```js
const y: number = 1 + +{};
```

It's a good way to get unexpected `NaN` and usually is a mistake.

**Comment by chicoxyzzy** [comment]
But `NaN` is a number so right part of your example is a valid expression.

**Comment by vkurchatkin** [comment]
I wouldn't call that explicit. It's the same as `'' + {}` for strings, which is not allowed.

**Comment by chicoxyzzy** [comment]
I'd rather say it's similar to
```
const y: number = +'1';
```
or
```
'' + {}.toString()
```
from your second example

**Comment by mroch** [comment]
i agree that it's sketchy, but since the only reason you would use the unary plus is to coerce the value (right?), we allow it. similarly, we allow `Number({})` because it's an explicit coercion.

**Comment by mroch** [comment]
I could see banning unary-plus on things that are guaranteed to give you `NaN`, but then that would require us to understand `valueOf`, which I don't think we get completely right currently.

**Comment by vkurchatkin** [comment]
Another twist: it throws on symbols the same way `'' + s` does

**Comment by chicoxyzzy** [comment]
This is because `+Symbol()` is a TypeError in JS
